docs/source/schema-directives.md

```diff
@@ -1,179 +1,527 @@
+
+* `visitSchema(schema: GraphQLSchema)`
+* `visitScalar(scalar: GraphQLScalarType)`
+* `visitObject(object: GraphQLObjectType)`
+* `visitFieldDefinition(field: GraphQLField<any, any>)`
+* `visitArgumentDefinition(argument: GraphQLArgument)`
+* `visitInterface(iface: GraphQLInterfaceType)`
+* `visitUnion(union: GraphQLUnionType)`
+* `visitEnum(type: GraphQLEnumType)`
+* `visitEnumValue(value: GraphQLEnumValue)`
+* `visitInputObject(object: GraphQLInputObjectType)`
+* `visitInputFieldDefinition(field: GraphQLInputField)`
+
+By overriding methods like `visitObject`, a subclass of `SchemaDirectiveVisitor` expresses interest in certain schema types such as `GraphQLObjectType` (the first parameter type of `visitObject`).
+
+When `SchemaDirectiveVisitor.visitSchemaDirectives` is called with a `GraphQLSchema` object and a map of visitor subclasses (`{ [directiveName: string]: typeof SchemaDirectiveVisitor }`), visitor methods overridden by those subclasses will be called with references to any schema type objects that have appropriately named `@directive`s attached to them, enabling the visitors to inspect or modify the schema.
+
+For example, if a directive called `@rest(url: "...")` appears after a field definition, a `SchemaDirectiveVisitor` subclass could provide meaning to that directive by overriding the `visitFieldDefinition` method (which receives a `GraphQLField` parameter), and then the body of that visitor method could manipulate the field's resolver function to fetch data from a REST endpoint:
+
+```typescript
+import {
+  makeExecutableSchema,
+  SchemaDirectiveVisitor,
+} from "graphql-tools";
 
-// Construct a schema, using GraphQL schema language
 const typeDefs = `
-  directive @upper on FIELD_DEFINITION
-
-  type Query {
-    hello: String @upper
-  }
-`;
-
-// Implement resolvers for out custom Directive
-const directiveResolvers = {
-  upper(
-    next,
-    src,
-    args,
-    context,
-  ) {
-    return next().then((str) => {
-      if (typeof(str) === 'string') {
-        return str.toUpperCase();
-      }
-      return str;
-    });
-  },
-}
+type Query {
+  people: [Person] @rest(url: "/api/v1/people")
+}`;
+
+const schema = makeExecutableSchema({ typeDefs });
+
+SchemaDirectiveVisitor.visitSchemaDirectives(schema, {
+  rest: class extends SchemaDirectiveVisitor {
+    public visitFieldDefinition(field: GraphQLField<any, any>) {
+      const { url } = this.args;
+      field.resolve = () => fetch(url);
+    }
+  }
+});
+```
 
-// Provide resolver functions for your schema fields
-const resolvers = {
-  Query: {
-    hello: (root, args, context) => {
-      return 'Hello world!';
-    },
-  },
-};
+The subclass in this example is defined as an anonymous `class` expression, for brevity. A truly reusable `SchemaDirectiveVisitor` would most likely be defined in a library using a named class declaration, and then exported for consumption by other modules and packages.
```

Comment:
I don't think documenting the anonymous class thing is worthwhile - I bet most people will find the existence of anonymous classes surprising

Reply:
Right, no need to mention it if the example isn't doing it anymore.